### DISCUSSION TITLE
Upgrade date-fns

### DISCUSSION THREAD

**Discussion by vykintazo** [rsuite/rsuite#3602] [General]
This library uses old `date-fns` v2 dependency, is there a reason it cannot be upgraded to the latest version?

**Comment by simonguo** [comment]
Hi @vykintazo , 
No reason, there are a lot of things to deal with, and upgrading `date-fns` has not been put on the agenda.

**Comment by simonguo** [comment]
This issue will be convert to discussion.

**Comment by vykintazo** [comment]
OK, that's understandable.
Originally I raised this question because I encountered a strange issue in my project. Now I had a bit more time to investigate it:
Something happened in rsuite update v5.49 that broke `date-fns` type resolutions for me. After looking through commits here It seems to be related to some imports.
My project uses `date-fns` v3 which has very different type definitions for most inputs than `date-fns` v2. And it worked all fine until rsuite v5.49. Now for some reason `tsc` command and my IDE started to resolve `date-fns` types from `node_modules/rsuite/date-fns` instead of  `node_modules/date-fns`. I couldn't override this in any way via `tsconfig.json`.
I am not entirely sure whose issue it is: rsuite, date-fns or typescript, but it did show up after rsuite upgrade.

The only solutions for now are either keep rsuite at v5.48 or completely disable types for `date-fns`, both of which are not ideal for me.
The third solution would be to upgrade date-fns to v3 in rsuite - I might be able to do a PR at some point, would that be alright?

**Reply by simonguo** [reply]
Can you send me the error logs or screenshots of the date-fns type definition error? Also, if upgrading the date-fns in rsuite to v3 does not cause any destructive changes, I think this PR is acceptable.

**Reply by vykintazo** [reply]
![image](https://github.com/rsuite/rsuite/assets/13200352/e443e031-e6c9-4092-b7c9-71d55bc307bc)
Note the cyan-colored paths to date-fns types. The main issue is that nearly all inputs for date-fns v3 functions accept `string` but in v2 they don't, so it yelds a lot of errors for me. Again, not entirely sure whether this is typescript or rsuite issue really, the only thing I know that until rsuite 5.49 everything was fine despite rsuite always using date-fns v2...

From what I've seen in this repo, upgrading date-fns to v3 would require to refactor nearly all imports from date-fns as they switched to tree-shakable named exports and rsuite heavily uses default exports with custom names sometimes. Other than that no functional changes would be necessary

**Comment by YochevedShapira** [comment]
I’m experiencing the same issue

**Comment by gagarwa** [comment]
We are still experiencing this issue.  It makes sense to roll back the version until this issue is resolved.
